File: src/commands/data/download.ts
import type { CommandoClient, CommandoMessage } from "discord.js-commando";

import { Command } from "discord.js-commando";
import * as embeds from "../../modules/embeds";
import { corpusManager } from "../../app";

import * as fs from "fs";
import * as FormData from "form-data";
import axios from "axios";


export default class DownloadCommand extends Command {
	constructor(client: CommandoClient) {
		super(client, {
			name: "download",
            memberName: "download",
            aliases: ["checkout"],
			group: "data",
            description: "Download a copy of your data.",
            throttling: {
                usages: 2,
                duration: 60 * 60, // 1 hour
            },
		});
    }
    

    async run(message: CommandoMessage): Promise<null> {
        const corpusFilePath = await corpusManager.pathTo(message.author);

        if (!corpusFilePath) {
            message.embed(embeds.noData(message.author));
            return null;
        }

        // Create an artificial HTML form to put the file in.
        // It's weird but that's how you POST files
        const form = new FormData();
        form.append("file", fs.createReadStream(corpusFilePath));

        // Upload to anonfiles because it has a free upload API lmao
        // TODO: Do anything but this
        const response = await axios.post("https://api.anonymousfiles.io/", form, {
            params: {
                expires: "6h",
                no_index: true,
            },
            headers: {
                ...form.getHeaders(),
            },
        });

        // Dig the new download URL out of the response data.
        const { url } = response.data;

        // DM the user their download link.
        await message.author.send(embeds.dataDownloadLink(url));

        // Tell them to check their DMs.
        message.embed(embeds.downloadReady);
        return null;
    }
};

File: commands/download.py
import aiohttp
from discord.ext import commands
from utils.pembed import Pembed
from utils.exceptions import NoDataError


class DownloadCommand(commands.Cog):
    def __init__(self, bot: commands.Bot) -> None:
        self.bot = bot

    @commands.command(aliases=["checkout"])
    @commands.cooldown(2, 3600, commands.BucketType.user)
    async def download(self, ctx: commands.Context) -> None:
        """ Download a copy of your data. """
        user = ctx.author

        try:
            corpus_file_path = self.bot.corpora.file_path(user)
        except FileNotFoundError:
            raise NoDataError(f"No data available for user {user.name}#{user.discriminator}")

        # Upload to file.io, a free filesharing service where the file is
        #   deleted once it's downloaded.
        with open(corpus_file_path, "rb") as f:
            async with aiohttp.ClientSession() as session:
                async with session.post("https://file.io/", data={"file": f, "expiry": "6h"}) as response:
                    download_url = (await response.json())["link"]

        # DM the user their download link.
        embed_download_link = Pembed(
            title="Link to download your data",
            description=download_url,
        )
        embed_download_link.set_footer(text="Link expires in 6 hours.")
        await user.send(embed=embed_download_link)

        # Tell them to check their DMs.
        embed_download_ready = Pembed(
            title="Download ready",
            color_name="green",
            description="A link to download your data has been DM'd to you.",
        )
        await ctx.send(embed=embed_download_ready)


def setup(bot: commands.Bot) -> None:
    bot.add_cog(DownloadCommand(bot))
